Symbol Table

CasperKjaerhus opened the pull request.

SoerenMLS committed:
CodeGen halløj

CodeGen halløj

SoerenMLS committed:
Code Gen Visitor revised

Revisited the visitor for code generation, made some new stuff, still alot of boring work need to be done :angry:

SoerenMLS committed:
Revisited pt2

he

SoerenMLS committed:
initial symboltable implementation

gots to be done by 22/04, lord i hope this is going to be true

SoerenMLS committed:
Symbol Table should be working.. ?

Symbol Table should be working..? not yet tested as we have no table-building yet :P.

SoerenMLS committed:
stufff stuff more stuff pls help

SoerenMLS committed:
SOMEWHAT WORKING SYMBOLTABLE IMPLEMENTATION, POG????

albania no.1 :100:

SoerenMLS committed:
Almost 100% working symboltable

Pog ?

SoerenMLS committed:
More stuff to the symbol table

assignments missing, also scoping needs a bit of tweaking :-)

SoerenMLS committed:
GetScopes

owo

SoerenMLS committed:
Now adds correct parent, and keeps track of depth

cum

anesheim-uni committed:
MonkW

SoerenMLS committed:
Plus Assignments done

need more assignments but else everything should be gucci

CasperKjaerhus committed:
Merge pull request #5 from CasperKjaerhus/CodeGen

Symbol Table to TypeChecker branch

SoerenMLS committed:
Get all symbols by their symbolname

CasperKjaerhus committed:
Types through AST Decoration

CasperKjaerhus committed:
Fix Speels hjerne

:)

SoerenMLS committed:
Cleaned up + Reference field added to vars

SoerenMLS committed:
More cleanup and refactoring

SoerenMLS committed:
Method that returns true if symbol is found in a scope

SoerenMLS committed:
some minor fixes

SoerenMLS committed:
Made the scope-list public for getting, and private for settingprivate

Long live OOP

SoerenMLS committed:
Removed an unnecessary class

CasperKjaerhus commented on Antlr/Test_Parser_src/Test_Parser/Tests/SymbolTableTest.txt:
```diff
@@ -0,0 +1,6 @@
+function main(number strLength) returns void begin
+
+number aaa;
+string poo;
```
💩 ?

anesheim-uni replied:
+1

CasperKjaerhus commented on Shell/SALShell/SALShell/Parser/ASTNode.cs:
```diff
@@ -1,5 +1,6 @@
 ﻿using System.Collections.Generic;
 using Antlr4.Runtime;
+using SALShell.CodeGen;
```
Why?

CasperKjaerhus commented on Shell/SALShell/SALShell/Parser/ASTVisitor.cs:
```diff
@@ -12,7 +12,7 @@ public virtual T Visit(ASTNode node)
                 return Visit((dynamic) child); // Using c#'s double dispatch feature
             }
-            return default(T);
+            return default;
```
Øøøh virker det her?

CasperKjaerhus commented on Shell/SALShell/SALShell/Parser/Nodes/FunctionDeclarationAstNode.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿using Antlr4.Runtime;
+using SALShell.CodeGen;
```
Why?

CasperKjaerhus commented on Shell/SALShell/SALShell/Parser/Nodes/ExprAstNode.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿using Antlr4.Runtime;
+using SALShell.CodeGen;
```
Why?

CasperKjaerhus commented on Shell/SALShell/SALShell/Parser/Nodes/ParameterListAstNode.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿using Antlr4.Runtime;
+using SALShell.CodeGen;
```
grr

CasperKjaerhus commented on Shell/SALShell/SALShell/Parser/Nodes/IdAstNode.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿using Antlr4.Runtime;
+using SALShell.CodeGen;
```
xd

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/DeclTypeInfo.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    class DeclTypeInfo : TypeInfo
+    {
+        IToken arraySize { get; }
```
Nok en god ide at omformatter det til et egentlig tal her?

leafantsky replied:
Slettes eller blive?

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/FuncCallTypeInfo.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.SymbolTable
+{
+    class FuncCallTypeInfo : TypeInfo
+    {
+        public FuncCallTypeInfo(IToken type) : base(type)
+        {
+        }
+
+        public override string ToString()
+        {
+            return $"Function Reference";
+        }
+    }
+}
```
Er function calls en del af symbol table?

SoerenMLS replied:
De bliver ikke gemt i symbol-tablet, men bliver brugt til at tjekke om der kan blive gjort kald til den refererede funktion i CheckFunctionReferences() metoden :^)

anesheim-uni commented on Shell/SALShell/SALShell/CodeGen/CodeEmitter.cs:
```diff
@@ -0,0 +1,121 @@
+        
+        private void ForeachEmitter(StreamWriter streamW, int numberOfElements, string ArrayName, string ForEachBody)
+        {
+            streamW.WriteLine($"for (int {numberOfElements} : {ArrayName})");
```
ligner ikke et ino eller c++ for loop?

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/Scope.cs:
```diff
@@ -0,0 +1,22 @@
+            Parent = parent;
+            this.scopeName = scopeName;
+            symbols = new List<Symbol>();
+            this.Depth = Depth; 
```
this. nogle steder og ikke andre steder?

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/Scope.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    class Scope
+    {
+        public Scope(Scope parent, string scopeName, int Depth)
```
Depth skal være med lille her

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableActual.cs:
```diff
@@ -0,0 +1,133 @@
+    class SymbolTableActual : ISymbolTable<Symbol, TypeInfo, Scope>
+    {
+
+        //THIS SYMBOL TABLE PROBABLY HAVE A VERY BAD SPACE AND TIME COMPLEXITY BUT ¯\_(ツ)_/¯
```
xd

leafantsky replied:
Nice troldpeter

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableActual.cs:
```diff
@@ -0,0 +1,133 @@
+
+namespace SALShell.SymbolTable
+{
+    class SymbolTableActual : ISymbolTable<Symbol, TypeInfo, Scope>
```
Virkelig dårlig navn til class ?

SoerenMLS replied:
Enig, det blev også done on a whim

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableActual.cs:
```diff
@@ -0,0 +1,133 @@
+                    Table.Add(prevSym.SymbolName, sym);
+                }
+            }
+            Depth--;
```
Comments til det her kunne nok være godt

SoerenMLS replied:
Outdated 👍

anesheim-uni reviewed (commented):
Meget enig med Cleth - efter request skal vi fikse type

leafantsky commented on Shell/SALShell/SALShell/SymbolTable/FuncTypeInfo.cs:
```diff
@@ -0,0 +1,27 @@
+
+        public override string ToString()
+        {
+            string funcString = "Function declaration taking the paremeters: ";
```
parameters*

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableActual.cs:
```diff
@@ -0,0 +1,133 @@
+        //Returns true if their exists a symbol with the name in the scope, in the given scope, else false.
+        public bool DeclaredInScope(string symbolName, string ScopeName)
+        {
+            return ScopeDisplay[ScopeDisplay.FindIndex(x => x.scopeName == ScopeName)].symbols.Any(x => x.SymbolName == symbolName);
```
Hvorfor ikke bare ScopeDisplay.find() i stedet for at først finde index, og så access igennem array?

SoerenMLS replied:
ved jeg faktisk ikke, er fixed nu :)

KasperLHP commented on Shell/SALShell/SALShell/SymbolTable/TypeInfo.cs:
```diff
@@ -0,0 +1,18 @@
+            Type = type;
+        }
+
+        public IToken Type { get; }
```
Fields efter constructor? kys

leafantsky commented on Shell/SALShell/SALShell/SymbolTable/IdTypeInfo.cs:
```diff
@@ -0,0 +1,37 @@
+                isReference = false;
+        }
+
+        IToken idName { get; }
```
field efter constructor smh

leafantsky commented on Shell/SALShell/SALShell/SymbolTable/Scope.cs:
```diff
@@ -0,0 +1,22 @@
+            this.Depth = Depth; 
+        }
+
+        public int Depth { get; }
```
field efter constructor

leafantsky commented on Shell/SALShell/SALShell/SymbolTable/Symbol.cs:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        public string scopeName { get; } 
+        public string SymbolName { get; }
```
xd placement

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableVisitor.cs:
```diff
@@ -0,0 +1,182 @@
+        {
+            IdAstNode idNode = (IdAstNode)node.Children[0];
+
+            return new AssignmentTypeInfo(idNode.Type);
```
Brug visitor i stedet for typecast

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableVisitor.cs:
```diff
@@ -0,0 +1,182 @@
+        public override TypeInfo Visit(DeclareAstNode node)
+        {
+            IdAstNode id = (IdAstNode)node.Id;
+            return new DeclTypeInfo(id.Type, id.ArraySize);
```
visitor i stedet for typecast

leafantsky commented on Shell/SALShell/SALShell/SymbolTable/Scope.cs:
```diff
@@ -0,0 +1,22 @@
+        public Scope(Scope parent, string scopeName, int Depth)
+        {
+            Parent = parent;
+            this.scopeName = scopeName;
```
sjovt mix

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableVisitor.cs:
```diff
@@ -0,0 +1,182 @@
+        public override TypeInfo Visit(FunctionDeclarationAstNode node)
+        {
+            IdAstNode IDinfo = (IdAstNode)node.Id;
+            ParameterListAstNode Paraminfo = (ParameterListAstNode)node.Parameters;
```
visitor i stedet for typecast

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/IdTypeInfo.cs:
```diff
@@ -0,0 +1,37 @@
+    class IdTypeInfo : TypeInfo
+    {
+        bool isArray { get; }
+        bool isReference { get; }
```
Skal symbol table holde styr på referencer?

SoerenMLS replied:
referencer bliver ikke gemt i symbol table

KasperLHP commented on Shell/SALShell/SALShell/SymbolTable/Scope.cs:
```diff
@@ -0,0 +1,22 @@
+        public Scope(Scope parent, string scopeName, int Depth)
+        {
+            Parent = parent;
+            this.scopeName = scopeName;
```
this. bruger du til at differentierer mellem constructor param og dit field fordi de har samme navn, men så gør du ikke det ved parent? burde man ikke stick til én convention (undskyld søren, ved ikke hvad jeg ellers skal finde)

CasperKjaerhus replied:
Tell him

SoerenMLS committed:
Reference checking :^) for symbols

SoerenMLS committed:
Checking for function references added

anesheim-uni committed:
Basic type checker

Implemented for:
Assignment, mult, plus, rel, equ, bool, if, while
Missing:
id, declaration, functioncall, arguments, functiondcl

SoerenMLS committed:
Cleaned up the code

anesheim-uni committed:
Merge pull request #7 from CasperKjaerhus/CodeGen

Code gen gib new symboltable

SoerenMLS committed:
Merge remote-tracking branch 'origin/TypeChecker' into CodeGen

SoerenMLS committed:
Merge remote-tracking branch 'origin/TypeChecker' into CodeGen

SoerenMLS committed:
Should now enrich the AST node with symbol

SoerenMLS committed:
Removed "Level" from Symbol Class

SoerenMLS committed:
updated some comments and ISymbolTable

CasperKjaerhus commented on Antlr/Test_Parser_src/Test_Parser/Tests/BoolsAndStrings.txt:
```diff
@@ -1,6 +1,8 @@
     string stringVar = "This is a string";
     
+    cum = 100;
+    
```
F

SoerenMLS replied:
also fixed

SoerenMLS committed:
Can add variables to the symbol table once againg

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/ISymbolTable.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace SALShell.SymbolTable
+{
+    interface ISymbolTable<T, K>
```
Måske en god kommentar her omkring hvad T og K er

SoerenMLS replied:
Done

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableActual.cs:
```diff
@@ -0,0 +1,172 @@
+
+namespace SALShell.SymbolTable
+{
+    class SymbolTableActual : ISymbolTable<Symbol, TypeInfo>
```
Generelt skal der flere kommentarer på symbol table. Den er ret 'magisk' fordi den springer frem og tilbage med hvor scopeindex og jeg ved ikke hvad bliver sat

CasperKjaerhus replied:
Navn på class skal ændres imo

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableBuilder.cs:
```diff
@@ -0,0 +1,91 @@
+        int forloopCount = 0;
+        int WhileCount = 0;
+        int ifCount = 0;
+        int switchCount = 0;
```
please brug private keyword

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableBuilder.cs:
```diff
@@ -0,0 +1,91 @@
+                    break;
+                case FunctioncallAstNode funcCallRef: 
+                    TypeInfo typeInfFuncCall = symVisitor.Visit(funcCallRef);
+                    IdAstNode Idinfo = (IdAstNode)funcCallRef.FunctionId;
```
type conversion av av av

SoerenMLS replied:
Visitor returner ikke det jeg skal bruge fra noden desværre :/

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableBuilder.cs:
```diff
@@ -0,0 +1,91 @@
+                    break;
+                case DeclareAstNode dclNode:
+                    TypeInfo typeInfoDcl = symVisitor.Visit(dclNode);
+                    IdAstNode dclId = (IdAstNode)dclNode.Id;
```
ooooooooooooooof

SoerenMLS replied:
Visitor returner ikke det jeg skal bruge fra noden desværre :/

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/SymbolTableVisitor.cs:
```diff
@@ -0,0 +1,182 @@
+
+        public override TypeInfo Visit(DeclareAstNode node)
+        {
+            IdAstNode id = (IdAstNode)node.Id;
```
Brug visitor i stedet for typecast

SoerenMLS replied:
Visitor returner ikke det jeg skal bruge fra noden desværre, så skulle jeg lave en ny visitor :/

SoerenMLS committed:
fixed some minor annoyances

SoerenMLS committed:
f

CasperKjaerhus commented on Antlr/Test_Parser_src/Test_Parser/Tests/BoolsAndStrings.txt:
```diff
@@ -1,6 +1,8 @@
-﻿function main(number strLength) returns void begin
+﻿unction main(number strLength) returns void begin
```
fix

CasperKjaerhus commented on Shell/SALShell/SALShell/SymbolTable/Symbol.cs:
```diff
@@ -0,0 +1,31 @@
+            scopeName = scopename;
+        }
+
+        public Symbol()
+        {
+        }
```
Er tom constructor needed?

SoerenMLS replied:
næh

CasperKjaerhus replied:
slet så

SoerenMLS committed:
Changed according to pull request comments

SoerenMLS committed:
Fixed the visitor, and the typecasts in symbol table visitor

SoerenMLS committed:
no more memes :(

SoerenMLS committed:
COMMENTS

SoerenMLS committed:
more comments and small fixes?
